project-base/storefront/components/Basic/UserText/GrapeJsParser.tsx

```diff
@@ -1,48 +1,60 @@
-import { ProductsSlider } from 'components/Blocks/Product/ProductsSlider';
-import { UserText } from 'components/Basic/UserText/UserText';
+import { ProductsList } from 'components/Blocks/Product/ProductsList/ProductsList';
 import { ListedProductFragmentApi, useProductsByCatnumsApi } from 'graphql/generated';
+import { GtmMessageOriginType, GtmProductListNameType } from 'gtm/types/enums';
+import { parseCatnums } from 'helpers/parsing/grapesJsParser';
 import { replaceAll } from 'helpers/replaceAll';
 import { memo } from 'react';
-import { GtmProductListNameType } from 'gtm/types/enums';
-import { GJS_PRODUCTS_SEPARATOR, parseCatnums } from 'helpers/parsing/grapesJsParser';
+import { UserText } from './UserText';
 
-type GrapesJsParserProps = {
+type GrapeJsParserProps = {
     text: string;
+    uuid: string;
 };
 
-export const GrapesJsParser: FC<GrapesJsParserProps> = memo(({ text }) => {
-    const productsCatnum = parseCatnums(text);
+export const GrapeJsParser: FC<GrapeJsParserProps> = memo(({ text, uuid }) => {
+    const catnums = parseCatnums(text);
+    const [allProductsResponse] = useProductsByCatnumsApi({ variables: { catnums } });
 
-    const dividedText = text.split(GJS_PRODUCTS_SEPARATOR).filter(Boolean);
-
-    const [result] = useProductsByCatnumsApi({ variables: { catnums: productsCatnum } });
+    const renderArticleProductList = (part: string) => {
+        if (!allProductsResponse.data?.productsByCatnums.length) {
+            return (
+                <ProductsList
+                    products={[]}
+                    gtmProductListName={GtmProductListNameType.other}
+                    gtmMessageOrigin={GtmMessageOriginType.other}
+                    key={uuid}
+                />
+            );
+        }
+
+        const products = replaceAll(part, /\[gjc-comp-ProductList&#61;|\]/g, '')
+            .split(',')
+            .map((product) =>
+                allProductsResponse.data?.productsByCatnums.find(
+                    (blogArticleProduct) => blogArticleProduct.catalogNumber === product,
+                ),
+            )
+            .filter(Boolean) as ListedProductFragmentApi[];
```

Comment:
Sometimes it can help to do type assertion, but in this case it seems to me like there is better assertion to do. Maybe better can be to put non-null assertion to the values which we are sure we have.

```suggestion
        const products = replaceAll(part, /\[gjc-comp-ProductList&#61;|\]/g, '')
            .split(',')
            .map((product) =>
                allProductsResponse.data!.productsByCatnums.find(
                    (blogArticleProduct) => blogArticleProduct.catalogNumber === product,
                )!,
            )
            .filter(Boolean);
```

This way you won't loose exact return type and only mark data which you will have for sure. `allProductsResponse.data` is checked in previous lines and `.find` will always find value since you are data are based on the same list.

Reply:
Without it, the code does not understand the type :cry: However, with a little rewrite, the assert is not necessary anymore.

Reply:
The way I wrote my suggestion it was working, but nevermind. Nice that it's working this way.